airbyte-integrations/connectors/source-shopify/source_shopify/source.py

```diff
@@ -0,0 +1,226 @@
+import math
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+import requests
+from airbyte_cdk import AirbyteLogger
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.models import SyncMode
+
+
+class ShopifyStream(HttpStream, ABC):
+    url_base = ""
+    primary_key = "id"
+    limit = 200
+
+    def __init__(self, shop: str, api_key: str, api_password: str, api_version: str, **kwargs):
+        super().__init__(**kwargs)
+        self.shop = shop
+        self.api_key = api_key
+        self.api_password = api_password
+        self.api_version = api_version
+        self.since_id = 0
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        decoded_response = response.json()
+        if len(decoded_response.get(self.data_field)) < self.limit:
+            return None
+        else:
+            self.since_id = decoded_response.get(self.data_field)[-1]["id"]
+            # UNCOMMENT FOR DEBUG ONLY
+            # print(f"\nSINCE_ID TOKEN: {self.since_id}\n")
+            # time.sleep(5)
+            return {"since_id": self.since_id}
+
+    def request_params(
+        self,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+        **kwargs
+        ) -> MutableMapping[str, Any]:
+        params = super().request_params(next_page_token=next_page_token, **kwargs)
+        params["limit"] = self.limit
+        params["since_id"] = self.since_id
+        if next_page_token:
+            params.update(**next_page_token)
+        return params
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        json_response = response.json()
+        records = json_response.get(self.data_field, []) if self.data_field is not None else json_response
+        for record in records:
+            yield record
+
+    @property
+    @abstractmethod
+    def data_field(self) -> str:
+        """The name of the field in the response which contains the data"""
+
+
+# Basic incremental stream
+class IncrementalShopifyStream(ShopifyStream, ABC):
+    state_checkpoint_interval = math.inf
+
+    @property
+    @abstractmethod
+    def cursor_field(self) -> str:
+        """
+        Defining a cursor field indicates that a stream is incremental, so any incremental stream must extend this class
+        and define a cursor field.
+        """
+        pass
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        return {self.cursor_field: max(latest_record.get(self.cursor_field), current_stream_state.get(self.cursor_field, 0))}
+
+### STREAM: customers
+class Customers(IncrementalShopifyStream):
+    data_field = "customers"
+    cursor_field = "id"
```

Comment:
why don't we use `updated_at` as the cursor field instead of ID? It's easier to understand. It seems like `Customer` and `AbandonedCheckouts` at least support this -- i haven't verified all endpoints but worth checking

Reply:
Based on Shopify's API Reference:
Ref link: [link](https://shopify.dev/docs/admin-api/rest/reference/orders/order)
"Filtering orders by updated_at is not an effective method for fetching orders because its value can change when no visible fields of an order have been updated. Use the Webhook and Event APIs to subscribe to order events instead."
They recommend using "since_id" because id is always "int", and it always += 1 for everything. Thus you can get the list of id's and sort them DESC, you'll get the latest first. So decided to use this. But probably, you've got the point here.